docs/layout-of-source-files.rst

```diff
@@ -216,101 +216,562 @@ the code below creates new global symbols ``alias`` and ``symbol2`` which refere
+
+Remapping allows you to use placeholders as import key prefixes and then have the compiler replace
+them with actual paths.
+For example you can set up a remapping so that everything imported from the virtual directory
+``github.com/ethereum/dapp-bin/library`` would actually be read from your local directory
+``/usr/local/dapp-bin/library``.
+
+The remappings can depend on a context, which allows you to configure packages to import,
+e.g. different versions of a library of the same name.
+
+.. warning::
+
+    Information about used remappings is stored in contract metadata so, while they let you avoid
+    changing the source, they cannot be used to achieve reproducible builds.
+    The metadata hash embedded in the bytecode will not be the same if you perform import remapping.
+
+Path remappings have the form of ``context:prefix=target``.
+All files in or below the ``context`` directory that import a file that starts with ``prefix`` are
+redirected by replacing ``prefix`` with ``target``.
+For example, if you clone ``github.com/ethereum/dapp-bin/`` locally to ``/usr/local/dapp-bin``,
+you can use the following in your source file:
 
 ::
 
-  import "github.com/ethereum/dapp-bin/library/iterable_mapping.sol" as it_mapping;
+    import "github.com/ethereum/dapp-bin/library/iterable_mapping.sol" as it_mapping;
 
 Then run the compiler:
 
 .. code-block:: bash
 
-  solc github.com/ethereum/dapp-bin/=/usr/local/dapp-bin/ source.sol
+    solc github.com/ethereum/dapp-bin/=/usr/local/dapp-bin/ source.sol
+
+As a more complex example, suppose you rely on a module that uses an old version of dapp-bin that
+you checked out to ``/usr/local/dapp-bin_old``, then you can run:
+
+.. code-block:: bash
+
+    solc module1:github.com/ethereum/dapp-bin/=/usr/local/dapp-bin/ \
+         module2:github.com/ethereum/dapp-bin/=/usr/local/dapp-bin_old/ \
+         source.sol
+
+This means that all imports in ``module2`` point to the old version but imports in ``module1``
+point to the new version.
+
+Here are the detailed rules governing the behavior of remappings:
+
+#. **Remappings only affect the translation between import paths and import keys.**
+
+   Import keys added via other means cannot be remapped.
+   For example the paths you specify on the command-line and the ones in ``urls`` in Standard JSON
+   are not affected.
+
+    .. code-block:: bash
+
+        solc /project=/contracts /project/contract.sol # Import key: /project/contract.sol
+
+#. **Context and prefix must match import keys, not import paths.**
+
+   - This means that you cannot remap ``./`` or ``../`` directly since they are replaced during
+     translation to import keys but you can remap the source locations they resolve into:
+
+     .. code-block:: bash
+
+         solc ./=a /project=b /project/contract.sol
+
+     .. code-block:: solidity
+         :caption: /project/contract.sol
+
+         import "./util.sol" as util; // Import key: b/util.sol
```

Comment:
I'm not convinced that being able to remap the resolved path is actually useful. See my comment from https://github.com/ethereum/solidity/issues/5281#issuecomment-790708752:

> Also, I wonder if the current behavior is actually useful in practice. In normal circumstances it seems redundant. For example if you do `/a/b=/x/y` then you change `import "/a/b/c.sol"` to `import "/x/y/c.sol"`. If `d.sol` then contains `import "./d.sol"`, the compiler will look for `d.sol` in `/x/y/` (because that's where `c.sol` is physically located) so the remapping does not play any role here.
>
> The only thing you can achieve by remapping the expanded path is the unintuitive behavior where `import "./d.sol"` ends up importing a file that's not in the same directory as `c.sol`. This would happen for example if you used `/a/b=/x/y /x/y=/a/b` in the example above.
>
> For that reason I think it would be better to just completely forbid remapping `.` and `..`. To me it looks like its main use would be writing underhanded code :)